### ISSUE TITLE
Date constructor should handle year 0 correctly

### ISSUE THREAD

**Issue by electrum** [prestodb/presto#5057]
Year zero is converted to January 2nd of 1 AD:

```
presto> select date '0000-12-31';
   _col0    
------------
 0001-01-02 
(1 row)
```

It should be rejected, similar to other invalid dates:
```
presto> select date '2015-02-29';
Query ... failed: Value cannot be cast to date: 2015-02-29

presto> select date '2016-12-32';
Query ... failed: Value cannot be cast to date: 2016-12-32

**Comment by nazgul33** [comment]
I didn't see you assigned yourself @geraint0923.
sorry about that and I'll cancel my PR if you want.

**Comment by martint** [comment]
0000 is a valid year. ANSI SQL uses the ISO 8601 notion of year, which allows anything between 0000 and 9999. 0000 is supposed to be 1 BC.

**Comment by geraint0923** [comment]
@nazgul33 You don't need to cancel your PR, you could continue to work on this. Thank you very much!

**Comment by nazgul33** [comment]
@martint then this is a totally different story. I think I need to find out what's wrong with joda time's iso dat formatter/parser logic.

**Comment by dain** [comment]
In my experience, Joda is always correct and my understanding of Date/Timestamps is wrong.

**Comment by geraint0923** [comment]
@nazgul33 I figured out that the problem is that we use joda which follows ISO standard and ignore the Julian calendar to parse but use the formatter in java.util which use both Julian calendar and Gregorian calendar to format the output string. I submitted a pull request #5073 in which we use Java 8 time API to format the output string only with Gregorian calendar. Thank you so much for your contribution!
